Add AnalyzerConfigOptionSet to get Option values from AnalyzerConfig

JoeRobich opened the pull request:
Currently most of the CodeStyle diagnostic analyzers rely on receiving a `WorkspaceAnalyzerOptions` (an Internal type) so that they can get a `OptionSet` to pull option values from. For projects that have .editorconfig files these option value can now be pulled from `AnalyzerConfigOptions` (a Public type).

This PR adds a AnalyzerConfigOptionSet that will first try to pull the Option value from `AnalyzerConfigOptions` and fall back to the document's `OptionSet` when no value is found. This benefits environments that have .editorconfig files and run analyzers, but do not have permissions to create instances of `WorkspaceAnalyzerOptions`.

sharwell commented on src/EditorFeatures/CSharpTest/Diagnostics/Configuration/ConfigureCodeStyle/BooleanCodeStyleOptionConfigurationTests.cs:
```diff
@@ -136,7 +136,7 @@ public Customer()
         </Document>
         <AnalyzerConfigDocument FilePath=""z:\\.editorconfig"">[*.cs]    # Comment1
-dotnet_style_object_initializer = false:suggestion    ; Comment2
+dotnet_style_object_initializer = true:suggestion    ; Comment2
```
Why was the test passing before this?

mavasani replied:
I think my analysis [below](https://github.com/dotnet/roslyn/pull/38675#discussion_r324007255) was not completely correct. I think previously the analyzer code to fetch option set returned null or default option value as the test diagnostic service does not use WorkspaceAnalyzerOptions. So analyzer flagged the diagnostic. The configuration code fix would still see the AnalyzerConfigDocument as the test workspace does have it, so it will work do the correct update to the editorconfig file.

jasonmalinowski commented on src/Workspaces/Core/Portable/CodeStyle/AnalyzerConfigOptionsExtensions.cs:
```diff
@@ -0,0 +1,63 @@
+{
+    internal static class AnalyzerConfigOptionsExtensions
+    {
+        private readonly static MethodInfo _tryGetEditorConfigOptionMethodInfo = typeof(AnalyzerConfigOptionsExtensions).GetMethod("TryGetEditorConfigOption", BindingFlags.NonPublic | BindingFlags.Static);
```
Yeah, this is the pattern I was expecting us to do.

mavasani replied:
Remove this unused private field now?

jasonmalinowski commented on src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace_XmlConsumption.cs:
```diff
@@ -280,7 +280,8 @@ public static TestWorkspace Create(string xmlDefinition, bool completed = true,
             else
             {
-                filePath = projectName +
+                // Prepend "./" because DocumentState expects the project to be within a directory when getting AnalyzerOptions.
```
This is correctly fixed in https://github.com/dotnet/roslyn/pull/39463/commits/2cbd65a304603cf897805a8888eed2c32cc427d3

JoeRobich replied:
reverted

JoeRobich commented:
Updated most usage to the API Manish described. I still need to update the apis where we pass an OptionSet out of the analyzer.

JoeRobich commented:
I've updated to match the api that @mavasani suggested. There are a few cases where the Analyzer invokes helper methods that are also used outside of Analyzer scenarios and that expect OptionsSet. So with @jasonmalinowski help I removed the reflection from the AnalyzerConfigOptionSet but left it for these cases. Ready for another round of review.

mavasani commented on src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs:
```diff
@@ -45,13 +45,7 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
-            }
-
-            var option = optionSet.GetOption(CSharpCodeStyleOptions.PreferBraces);
+            var option = context.Options.GetOption(CSharpCodeStyleOptions.PreferBraces, statement.SyntaxTree, cancellationToken);
```
This is nice! We can potentially go a step further and define extension methods on common context types and get the tree and cancellationToken from the context. For example:

```csharp
T GetOption<T>(this SyntaxNodeAnalysisContext context, IOption<T> option);
T GetOption<T>(this OperationAnalysisContext context, IOption<T> option);

mavasani commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypeStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -48,11 +48,7 @@ private void HandleVariableDeclaration(SyntaxNodeAnalysisContext context)
-            {
-                return;
-            }
+            var optionSet = options.GetAnalyzerOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
```
It seems that all the call sites for this method have a `.GetAwaiter().GetResult()`. Might be useful to define a `GetAnalyzerOptionSet` helper which does this.

jasonmalinowski replied:
Agreed, and especially because once we rip out an remaining dependencies on fetching the Workspace under the covers the async-ness would go away anyways. Better to do it once now.

mavasani commented on src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -149,8 +149,7 @@ void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
 
-        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
-        private static async ValueTask<NamingStylePreferences> GetNamingStylePreferencesAsync(
+        private static NamingStylePreferences GetNamingStylePreferencesAsync(
```
Nit: rename to `GetNamingStylePreferences`. Was the removal of `PerformanceSensitiveAttribute` intentional? Tagging @sharwell

JoeRobich replied:
It was not. I'll revert.

mavasani commented on src/Features/Core/Portable/Formatting/FormattingDiagnosticAnalyzer.cs:
```diff
@@ -35,11 +35,6 @@ private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
             var cancellationToken = context.CancellationToken;
 
             var options = context.Options.GetDocumentOptionSetAsync(tree, cancellationToken).GetAwaiter().GetResult();
```
We are not moving this analyzer to use `GetOption` API?

JoeRobich replied:
This analyzer has a hard requirement on `Workspace`, but I will update it to use `GetAnalyzerOptionSet` which is a step in the right direction.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/AnalyzerConfigOptionSet.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal class AnalyzerConfigOptionSet : OptionSet
```
Can you add a doc comment of some kind explaining what this one is?

jasonmalinowski commented on src/Workspaces/Core/Portable/CodeStyle/AnalyzerConfigOptionsExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+            return TryGetEditorConfigOption(analyzerConfigOptions, option, out value);
+        }
+
+        internal static bool TryGetOption(this AnalyzerConfigOptions analyzerConfigOptions, OptionKey optionKey, out object value)
```
What is this needed for? I would have assumed everybody was using the slightly more strongly typed form below?

jasonmalinowski commented on src/Workspaces/Core/Portable/Options/EditorConfig/EditorConfigStorageLocationExtensions.cs:
```diff
@@ -0,0 +1,66 @@
+            return editorConfigStorageLocation.TryGetOption(new AnalyzerConfigOptionsDictionary(analyzerConfigOptions), type, out value);
+        }
+
+        private class AnalyzerConfigOptionsDictionary : IReadOnlyDictionary<string, string>
```
Add a comment this is just wrapping an AnalyzerConfigOptions into an interface.

jasonmalinowski commented on src/Workspaces/Core/Portable/CodeStyle/AnalyzerConfigOptionsExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+        {
+            foreach (var storageLocation in option.StorageLocations)
+            {
+                if (storageLocation is EditorConfigStorageLocation<T> editorConfigStorageLocation &&
```
I would have thought this would be implemented like the version above where you cast to just IEditorConfigStorageLocation, call the extension there, and then when you get the resulting object just cast it to T.

JoeRobich replied:
Talked offline and consolidated the implementations

jasonmalinowski commented on src/Workspaces/Core/Portable/CodeStyle/AnalyzerConfigOptionsExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+
+        private static bool TryGetEditorConfigOption<T>(this AnalyzerConfigOptions analyzerConfigOptions, IOption option, out T value)
+        {
+            foreach (var storageLocation in option.StorageLocations)
```
Should this throw if none of the storage locations were .editor config storage locations? Mostly because it seems like that would be an error that something isn't hooked up right?

JoeRobich replied:
Added throw.

jasonmalinowski replied:
Not sure I see the throw...?

JoeRobich replied:
Sorry added then later removed throw. This is called by the AnalyzerConfigOptionsSet which is passed into code that expects a full OptionsSet. I could open a new issue to add EditorConfigStorage to all the options that are being used in this manner.

JoeRobich replied:
Opened https://github.com/dotnet/roslyn/issues/40464 to track this.

jasonmalinowski commented on src/Features/CSharp/Portable/UseDefaultLiteral/CSharpUseDefaultLiteralCodeFixProvider.cs:
```diff
@@ -52,6 +53,7 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
 
             var parseOptions = (CSharpParseOptions)document.Project.ParseOptions;
             var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var preferSimpleDefaultExpression = options.GetOption(CSharpCodeStyleOptions.PreferSimpleDefaultExpression).Value;
```
Is the .Value earlier now safe? Previously it wouldn't have ran unless something else happened earlier?

jasonmalinowski commented on src/Workspaces/Core/Portable/CodeStyle/AnalyzerConfigOptionsExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+        private readonly static MethodInfo _tryGetEditorConfigOptionMethodInfo = typeof(AnalyzerConfigOptionsExtensions).GetMethod("TryGetEditorConfigOption", BindingFlags.NonPublic | BindingFlags.Static);
+
+        public static bool TryGetOption<T>(this AnalyzerConfigOptions analyzerConfigOptions, Option<T> option, out T value)
```
Can we rename these TryGetEditorConfigOption as well? That way it's clear this fetches just the editorconfig option?

jasonmalinowski replied:
(the specific confusion being there's a lot of other calls to this and it's sometimes unclear what *those* are calling...)

jasonmalinowski replied:
Doesn't look like this was applied?

JoeRobich replied:
Good catch! =)

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs:
```diff
@@ -104,6 +105,52 @@ public static (string analyzerId, VersionStamp version) GetAnalyzerIdAndVersion(
+            return GetOptionAsync(analyzerOptions, option, language, syntaxTree, cancellationToken).GetAwaiter().GetResult();
+        }
+
+        public static async ValueTask<T> GetOptionAsync<T>(this AnalyzerOptions analyzerOptions, PerLanguageOption<T> option, string language, SyntaxTree syntaxTree, CancellationToken cancellationToken)
```
Can't this and the the GetOptionAsync<T>() that takes an `Option<T>` just call into a shared implementation that takes an OptionKey?

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs:
```diff
@@ -104,6 +105,52 @@ public static (string analyzerId, VersionStamp version) GetAnalyzerIdAndVersion(
+        }
+
         [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
         public static ValueTask<OptionSet?> GetDocumentOptionSetAsync(this AnalyzerOptions analyzerOptions, SyntaxTree syntaxTree, CancellationToken cancellationToken)
```
Can we now mark this [Obsolete]? Other than the uses in this class, and the few cases where we have APIs that still take a full option set and are thus tied to Workspace specific stuff, this shouldn't be used anymore? I wonder if it's worth marking it [Obsolete] so we don't get new dependencies on it.

JoeRobich replied:
Marked as obsolete.

JoeRobich commented:
@mavasani & @jasonmalinowski  Updated based on your feedback which caused a bit of churn. Please review when you have time.

mavasani commented on src/Features/Core/Portable/PreferFrameworkType/PreferFrameworkTypeDiagnosticAnalyzerBase.cs:
```diff
@@ -50,18 +51,30 @@ public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
         protected sealed override void InitializeWorker(AnalysisContext context)
             => context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKindsOfInterest);
 
+        void Test()
```
Revert?

JoeRobich replied:
This is why I need a 3rd Roslyn enlistment, so I don't test things in the same place I am making changes. Will revert.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -151,7 +151,7 @@ void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
         }
 
         [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
```
This attribute can be removed now.

jasonmalinowski commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerConfigOptions.cs:
```diff
@@ -23,18 +22,27 @@ public abstract class AnalyzerConfigOptions
 
         public override bool TryGetValue(string key, out string value) => _backing.TryGetValue(key, out value);
 
+        public IReadOnlyDictionary<string, string> AsReadOnlyDictionary() => _backing;
```
@agocke Didn't we want to avoid exposing this as a IReadOnlyDictionary for the most part? Because now this exposes enumeration which is Complicated™?

jasonmalinowski replied:
It also seems that if we _did_ want to encode an API for this, it should be an interface, so other implementations could expose this without forcing their inheritence hierarchy.

JoeRobich replied:
Getting NamingStyleOptions expects the ability to enumerate all the keys defined in the .editorconfig. My dictionary wrapper for AnalyzerConfigOptions was insufficient since I was unable to provide that capability.

JoeRobich replied:
@agocke @sharwell Do either of you remember discussing how to account for NamingStyleOptions when designing the .editorconfig support?

JoeRobich replied:
After talking to @agocke, an enumerator was not offered because the expectation that the IDE would implement the AnalyzerConfigOptions in a way that merged the config options with the Workspace OptionSet. At the time, the OptionSet was not part of the Solution snapshot so creating a reliable enumeration wasn't possible. Now that OptionSet is tied to a snapshot we should be able to enumerate all the options without fear of things changing underneath us.

jasonmalinowski commented on src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs:
```diff
@@ -4,29 +4,29 @@
         }
 
-        private static bool TryGetEditorConfigOption<T>(this OptionSet analyzerConfigOptions, IOption option, out T value)
+        private static bool TryGetEditorConfigOption<T>(this OptionSet optionSet, IOption option, out T value)
```
What does this method do that optionSet.GetOption<T> doesn't do directly? The walking to find the explicit EditorConfigStorageLocation omits the naming styles case...?

jasonmalinowski replied:
(maybe my more general question is "who uses these methods?"?)

JoeRobich replied:
Looks like most things in the `*.CodeStyle/Formatting` use these methods.

JoeRobich replied:
Many files are shared between the Workspace and CodeStyle layers. These extension methods are to handle getting option values when being built as part of the CodeStyle layer and need to match the method names used in the Workspace layer.

jasonmalinowski commented on src/CodeStyle/Core/Analyzers/Options/OptionSetExtensions.cs:
```diff
@@ -4,29 +4,29 @@
+    internal static class OptionSetExtensions
     {
-        public static T GetOption<T>(this OptionSet analyzerConfigOptions, Option<T> option)
+        public static T GetOption<T>(this OptionSet optionSet, Option<T> option)
```
Should these be renamed? Something like GetEditorConfigOptionByOnlyReadingEditorConfig or something?

JoeRobich replied:
How is `GetOptionFromEditorConfig` ?

JoeRobich replied:
These cannot be renamed as they are the codestyle layers implementation of extension methods that exist in the workspace layer. They are here to allow code that is linked into both codestyle and workspace layers to build properly.

jasonmalinowski commented on src/Features/Core/Portable/Diagnostics/AnalyzerConfigOptionSet.cs:
```diff
@@ -0,0 +1,47 @@
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// This class proxies requests for Option values first to the AnalyzerConfigOptions then to a backup OptionSet if provided.
```
Use cref.

jasonmalinowski commented on src/Features/Core/Portable/UseSystemHashCode/UseSystemHashCodeDiagnosticAnalyzer.cs:
```diff
@@ -49,14 +49,10 @@ private void AnalyzeOperationBlock(Analyzer analyzer, OperationBlockAnalysisCont
-            }
 
-            var option = optionSet.GetOption(CodeStyleOptions.PreferSystemHashCode, operation.Language);
+            var option = context.Options.GetOption(CodeStyleOptions.PreferSystemHashCode, operation.Language, syntaxTree, cancellationToken);
```
No fancy extension method for this one? Fine if there isn't, just not sure if this got missed.

JoeRobich replied:
No fancy one for this one. Not a lot of Operation Block Analyzers in Roslyn.

CyrusNajmabadi commented:
Can we prioritize this goign in?  It's a major part of https://github.com/dotnet/roslyn/pull/41363 (which is gargantuan and could use some love).

JoeRobich commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerConfigOptions.cs:
```diff
@@ -23,6 +22,11 @@ public abstract class AnalyzerConfigOptions
+        /// <summary>
+        /// Get the keys of the defined options.
+        /// </summary>
+        public abstract ImmutableArray<string> Keys { get; }
```
@jasonmalinowski @agocke Does this seem like a workable solution to enumerating the options? Now that the OptionSet is part of the solution snapshot. We can work on combining the solution options with the analyzer options in a follow up PR.

JoeRobich replied:
After talking offline to Andy, I updated to `IEnumerable<string> Keys`.

sharwell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpSimplifyTypeNamesDiagnosticAnalyzer.cs:
```diff
@@ -52,7 +52,9 @@ protected override void AnalyzeCodeBlock(CodeBlockAnalysisContext context)
             var options = context.Options;
+#pragma warning disable CS0612 // Type or member is obsolete
             var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult()!;
+#pragma warning restore CS0612 // Type or member is obsolete
```
❔ Why are these not changing?

JoeRobich replied:
Thank you for pointing this out. I mindlessly fixed it up when resolving merge conflicts.

sharwell commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypeStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -50,11 +50,7 @@ private void HandleVariableDeclaration(SyntaxNodeAnalysisContext context)
-            {
-                return;
-            }
+            var optionSet = options.GetAnalyzerOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
```
😕 Why are we not using `context.Options` directly? This option set should automatically contain the options from the IDE when invoked as a VSIX-installed analyzer, and omit those options when called via a NuGet-installed analyzer.

JoeRobich replied:
I don't think I am following. The AnalyzerOptionSet provides the behavior you are describing. Are you saying the AnalyzerConfigOptions provides this behavior without needing to wrap the DocumentOptionSet?

agocke commented:
Why was this merged?
